keps/sig-scheduling/20190221-even-pods-spreading.md

```diff
@@ -0,0 +1,415 @@
+option to trigger "toleration" mode when the cluster is out of resource. Then in
+aforementioned example, a third pod is "tolerated" to be put onto node1 (or
+node2). But keep it in mind, this behavior is only triggered upon resource
+shortage. For a 3-nodes cluster, the third pod will still be placed onto node3
+(if node3 is capable).
+
+### Risks and Mitigations
+
+The feature requires additional processing for pods that use it and it is ok to
+have some performance overhead. But we will make sure our implementation will
+not have any performance penalty for pods that do not use this feature.
+
+## Design Details
+
+Basically there are two ways in API design and implementation:
+
+1. implemented as a "sub feature" inside Affinity and hence new API applies to
+   `pod.spec.affinity`. (referred as [Design A](#design-b) in this doc)
+1. implemented as a standalone feature and new API applies to `pod.spec`.
+   (referred as [Design B](#design-b) in this doc)
+
+### Design A
+
+A new structure called `EvenSpreading` is introduced and it's only effective
+when it's not nil.
+
+```go
+type EvenSpreading struct {
+    // MaxSkew describes the degree of imbalance of pods spreading.
+    // Default value is 1 and 0 is not allowed.
+    MaxSkew int32
+    // TopologyKey defines where pods are placed evenly
+    TopologyKey string
+}
+```
+
+In this design, `EvenSpreading` is a field in Affinity specs:
+
+```go
+type NodeAffinity struct {
+    EvenSpreading *EvenSpreading
+    ......
+}
+
+type PodAffinity struct {
+    EvenSpreading *EvenSpreading
+    ......
+}
+
+type PodAntiAffinity struct {
+    EvenSpreading *EvenSpreading
+    ......
+}
+```
+
+### Design B
+
+Unlike Design A, `EvenSpreading` acts as a standalone spec and applies to
+`pod.spec`. Similarly it's only effective when it's not nil.
+
+```go
+type PodSpec struct {
+    EvenSpreading *EvenSpreading
+    ......
+}
+```
+
+Inside `EvenSpreading`, we need hard affinityTerms (similar with
+`PodAffinityTerm`) and soft affinityTerms (similar with
+`WeightedPodAffinityTerm`). This describes when we perform even distribution,
+which pods are considered as a group.
+
+```go
+type EvenSpreading struct {
+    // MaxSkew describes the degree of imbalance of pods spreading.
+    // Default value is 1 and 0 is not allowed.
+    MaxSkew int32
+    // TopologyKey defines where pods are placed evenly
+    TopologyKey string
+    // Similar with the same field in PodAffinity/PodAntiAffinity
+    // +optional
+    RequiredDuringSchedulingIgnoredDuringExecution []PodAffinityTerm
+    // Similar with the same field in PodAffinity/PodAntiAffinity
+    // +optional
+    PreferredDuringSchedulingIgnoredDuringExecution []WeightedPodAffinityTerm
+}
+```
+
+### MaxSkew
+
+`MaxSkew` and the formula (`matching num - minMatching num < MaxSkew`) is core
+of this KEP, so the exact semantics are clarified as below:
+
+- chicken/egg problem
+
+Let's say we have a 3-zone cluster, and there is no pod in any node yet. Here
```

Comment:
With the changes I mentioned above this problem won't exists at all. You will have a pod selector that initially does not match any pods in the cluster. MaxSkew is 1. If you schedule a new pod in any of the three zones, MaxSkew won't get violated. So, all the zones will be feasible for the first pod and you won't need any extra checks.

Reply:
Yeap.

Reply:
If we use a node selector, then yes.